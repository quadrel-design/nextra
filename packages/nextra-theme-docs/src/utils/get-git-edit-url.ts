import { useConfig } from '../contexts'
import parseGitUrl from 'git-url-parse'

export const getGitEditUrl = (filePath?: string): string => {
  const config = useConfig()
  const repo = parseGitUrl(config.docsRepositoryBase || '')

  if (!repo) throw new Error('Invalid `docsRepositoryBase` URL!')

  return `${repo.href}/${filePath}`
}