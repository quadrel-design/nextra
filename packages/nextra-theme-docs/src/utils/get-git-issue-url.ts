import gitUrlParse from 'git-url-parse'

export const getGitIssueUrl = ({
  repository = '',
  title,
  labels
}: {
  repository?: string
  title: string
  labels?: string
}): string => {
  const repo = gitUrlParse(repository)
  if (!repo) throw new Error('Invalid `docsRepositoryBase` URL!')

  if (repo.resource.includes('gitlab')) {
    return `https://gitlab.com/${repo.owner}/${
      repo.name
    }/-/issues/new?issue[title]=${encodeURIComponent(title)}`
  } else if (repo.resource.includes('github')) {
    return `https://github.com/${repo.owner}/${
      repo.name
    }/issues/new?title=${encodeURIComponent(title)}&labels=${labels || ''}`
  } else {
    return '#'
  }
}
