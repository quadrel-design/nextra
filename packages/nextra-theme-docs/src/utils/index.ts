'use no memo'

export { getGitIssueUrl } from './get-git-issue-url'
export { gitUrlParse } from './git-url-parse'
