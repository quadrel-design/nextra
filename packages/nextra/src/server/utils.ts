import path from 'path'
import type {
  ArrayExpression,
  ExportNamedDeclaration,
  Expression,
  ObjectExpression
} from 'estree'
import slash from 'slash'
import title from 'title'
import type { Folder, MdxFile } from '../types'
import { DEFAULT_PROPERTY_PROPS } from './constants.js'

type Truthy<T> = T extends false | '' | 0 | null | undefined ? never : T // from lodash

export function truthy<T>(value: T): value is Truthy<T> {
  return !!value
}

export const logger = {
  info: console.log.bind(null, '-', '\x1b[36minfo\x1b[0m', '[nextra]'),
  warn: console.log.bind(null, '-', '\x1b[33mwarn\x1b[0m', '[nextra]'),
  error: console.log.bind(null, '-', '\x1b[31merror\x1b[0m', '[nextra]')
}

export function pageTitleFromFilename(fileName: string) {
  return title(fileName.replaceAll(/[-_]/g, ' '), { special: ['SSR'] })
}

export function sortPages(
  pages: (Omit<MdxFile, 'route'> | Omit<Folder, 'route' | 'children'>)[],
  locale?: string
): [string, string][] {
  return pages
    .map(item => ({
      name: item.name,
      date: 'frontMatter' in item && item.frontMatter?.date,
      title:
        ('frontMatter' in item && item.frontMatter?.title) ||
        pageTitleFromFilename(item.name)
    }))
    .sort((a, b) => {
      if (a.date && b.date) {
        return new Date(b.date).getTime() - new Date(a.date).getTime()
      }
      if (a.date) {
        return -1 // sort a before b
      }
      if (b.date) {
        return 1 // sort a after b
      }
      return a.title.localeCompare(b.title, locale || undefined, {
        numeric: true
      })
    })
    .map(item => [item.name, item.title])
}

export function normalizePageRoute(parentRoute: string, route: string): string {
  return slash(path.join(parentRoute, route.replace(/^index$/, '')))
}

export function createAstExportConst(
  name: string,
  value: ArrayExpression | ObjectExpression | Expression
): ExportNamedDeclaration {
  return {
    type: 'ExportNamedDeclaration',
    specifiers: [],
    declaration: {
      type: 'VariableDeclaration',
      kind: 'const',
      declarations: [
        {
          type: 'VariableDeclarator',
          id: { type: 'Identifier', name },
          init: value
        }
      ]
    }
  }
}

export function createAstObject(
  obj: Record<string, string | number | boolean | null | Expression>
): ObjectExpression {
  return {
    type: 'ObjectExpression',
    properties: Object.entries(obj).map(([key, value]) => ({
      ...DEFAULT_PROPERTY_PROPS,
      key: { type: 'Identifier', name: key },
      value:
        value && typeof value === 'object' ? value : { type: 'Literal', value }
    }))
  }
}
