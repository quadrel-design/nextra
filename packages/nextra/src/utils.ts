import path from 'node:path'
import slash from 'slash'
import title from 'title'
import type { Folder, MdxFile, Meta } from './types'

type Truthy<T> = T extends false | '' | 0 | null | undefined ? never : T // from lodash

export function truthy<T>(value: T): value is Truthy<T> {
  return !!value
}

export function normalizeMeta(meta: Meta): Exclude<Meta, string> {
  return typeof meta === 'string' ? { title: meta } : meta
}

export function normalizePageRoute(parentRoute: string, route: string): string {
  return slash(path.join(parentRoute, route.replace(/^index$/, '')))
}

export function pageTitleFromFilename(fileName: string) {
  return title(fileName.replaceAll(/[-_]/g, ' '))
}

export function sortPages(
  pages: (
    | Pick<MdxFile, 'kind' | 'name' | 'frontMatter' | 'locale'>
    | Pick<Folder, 'kind' | 'name'>
  )[],
  locale?: string
): [string, string][] {
  if (locale === '') {
    locale = undefined
  }
  return pages
    .filter(item => item.kind === 'Folder' || item.locale === locale)
    .map(item => ({
      name: item.name,
      date: 'frontMatter' in item && item.frontMatter?.date,
      title:
        ('frontMatter' in item && item.frontMatter?.title) ||
        pageTitleFromFilename(item.name)
    }))
    .sort((a, b) => {
      if (a.date && b.date) {
        return new Date(b.date).getTime() - new Date(a.date).getTime()
      }
      if (a.date) {
        return -1 // sort a before b
      }
      if (b.date) {
        return 1 // sort a after b
      }
      return a.title.localeCompare(b.title, locale, { numeric: true })
    })
    .map(item => [item.name, item.title])
}

export function isSerializable(o: any): boolean {
  try {
    JSON.stringify(o)
    return true
  } catch (err) {
    return false
  }
}

/**
 * Calculate a 32 bit FNV-1a hash
 * Found here: https://gist.github.com/vaiorabbit/5657561
 * Ref.: http://isthe.com/chongo/tech/comp/fnv/
 *
 * @param {string} str the input value
 * @param {number} [seed] optionally pass the hash of the previous chunk
 * @returns {string}
 */
export function hashFnv32a(str: string, seed = 0x811c9dc5): string {
  let hval = seed

  for (let i = 0; i < str.length; i++) {
    hval ^= str.charCodeAt(i)
    hval += (hval << 1) + (hval << 4) + (hval << 7) + (hval << 8) + (hval << 24)
  }

  // Convert to 8 digit hex string
  return ('0000000' + (hval >>> 0).toString(16)).substring(-8)
}

export function getDefault<T>(module: T & { default?: T }): T {
  return module.default || module
}

export const logger = {
  info: console.log.bind(null, '- ', '\x1b[36minfo\x1b[0m', '[nextra]'),
  warn: console.log.bind(null, '- ', '\x1b[33mwarn\x1b[0m', '[nextra]'),
  error: console.log.bind(null, '- ', '\x1b[31mwarn\x1b[0m', '[nextra]')
}
