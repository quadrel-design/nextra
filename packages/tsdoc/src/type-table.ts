import type { GenerateDocumentationOptions } from './base.js'
import { generateDocumentation } from './base.js'

export interface BaseTypeTableProps {
  /**
   * Set the type to generate from.
   */
  code: string

  options?: GenerateDocumentationOptions & {
    /**
     * base path to resolve `path` prop
     */
    basePath?: string
  }
}

export async function getTypeTableOutput({
  code,
  options
}: BaseTypeTableProps) {
  const output = generateDocumentation('temp.ts', 'default', code, options)
  if (!output.length) {
    throw new Error("Can't find `export default` statement")
  }
  return output
}
