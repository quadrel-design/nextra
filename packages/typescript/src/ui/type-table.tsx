import cn from 'clsx'
import Link from 'next/link'
import { Code, Popup } from 'nextra/components'
import { InformationCircleIcon as InfoIcon } from 'nextra/icons'
import type { FC, ReactNode } from 'react'

const Info: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <Popup>
      <Popup.Button>
        <InfoIcon className="x:size-4" />
      </Popup.Button>
      <Popup.Panel className="x:max-h-[400px] x:max-w-[400px] x:min-w-[220px] x:overflow-auto x:text-sm">
        {children}
      </Popup.Panel>
    </Popup>
  )
}

interface ObjectType {
  /**
   * Additional description of the field
   */
  description: ReactNode
  type: string
  typeDescription: ReactNode
  /**
   * Optional link to the type
   */
  typeDescriptionLink?: string
  default?: string
  required?: boolean
}

const classes = {
  field: cn('x:inline-flex x:items-center x:gap-1'),
  code: cn('x:rounded-md bg-fd-secondary x:p-1 text-fd-secondary-foreground'),
  blue: 'bg-fd-primary/10 text-fd-primary'
}

export const TypeTable: FC<{ type: Record<string, ObjectType> }> = ({
  type
}) => {
  return (
    <div className="x:my-6 x:overflow-x-auto nextra-scrollbar">
      <table className="text-fd-muted-foreground x:text-sm">
        <thead>
          <tr>
            <th align="left" className="x:w-1/3">
              Name
            </th>
            <th align="left">Type</th>
            <th align="left">Default</th>
          </tr>
        </thead>
        <tbody>
          {Object.entries(type).map(([key, value]) => (
            <tr key={key}>
              <td>
                <div className={classes.field}>
                  <Code>
                    {key}
                    {!value.required && '?'}
                  </Code>
                  {value.description && <Info>{value.description}</Info>}
                </div>
              </td>
              <td>
                <Code>{value.type}</Code>
                {value.typeDescription && <Info>{value.typeDescription}</Info>}
                {value.typeDescriptionLink && (
                  <Link href={value.typeDescriptionLink}>
                    <InfoIcon className="x:size-4" />
                  </Link>
                )}
              </td>
              <td>
                {value.default ? <Code>{value.default}</Code> : <span>-</span>}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  )
}
