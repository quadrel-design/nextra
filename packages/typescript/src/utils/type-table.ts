import fs from 'node:fs/promises'
import path from 'node:path'
import type { GenerateDocumentationOptions } from '../lib/base.js'
import { generateDocumentation } from '../lib/base.js'

export interface BaseTypeTableProps {
  /**
   * The path to source TypeScript file.
   */
  path?: string

  /**
   * Exported type name to generate from.
   */
  name?: string

  /**
   * Set the type to generate from.
   *
   * When used with `name`, it generates the type with `name` as export name.
   *
   * ```ts
   * export const myName = MyType;
   * ```
   *
   * When `type` contains multiple lines, `export const` is not added.
   * You need to export it manually, and specify the type name with `name`.
   *
   * ```tsx
   * <AutoTypeTable
   *   path="./file.ts"
   *   type={`import { ReactNode } from "react"
   *   export const MyName = ReactNode`}
   *   name="MyName"
   * />
   * ```
   */
  type?: string

  options?: GenerateDocumentationOptions & {
    /**
     * base path to resolve `path` prop
     */
    basePath?: string
  }
}

export async function getTypeTableOutput({
  name,
  type,
  options,
  ...props
}: BaseTypeTableProps) {
  const file =
    props.path && options?.basePath
      ? path.join(options.basePath, props.path)
      : props.path
  let typeName = name
  let content = ''

  if (file) {
    content = await fs.readFile(file, 'utf8')
  }

  if (type && type.split('\n').length > 1) {
    content += `\n${type}`
  } else if (type) {
    typeName ??= '$Fumadocs'
    content += `\nexport type ${typeName} = ${type}`
  }

  const output = generateDocumentation(
    file ?? 'temp.ts',
    typeName,
    content,
    options
  )

  if (name && output.length === 0)
    throw new Error(`${name} in ${file ?? 'empty file'} doesn't exist`)

  return output
}
