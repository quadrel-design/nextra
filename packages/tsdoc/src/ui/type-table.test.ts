import { getTypeTableOutput } from '../utils/type-table'

describe('TypeTable', () => {
  it('<Banner />', async () => {
    const code = `import type { Banner } from 'nextra/components'
type $ = React.ComponentProps<typeof Banner>
export default $`
    const result = await getTypeTableOutput({ code })
    expect(result).toMatchSnapshot()
  })
  it('<Search />', async () => {
    const code = `import type { Search } from 'nextra/components'
type $ = React.ComponentProps<typeof Search>
export default $`
    const result = await getTypeTableOutput({ code })
    expect(result).toMatchSnapshot()
  })
  it('<Callout />', async () => {
    const code = `import type { Callout } from 'nextra/components'
type $ = React.ComponentProps<typeof Callout>
export default $`
    const result = await getTypeTableOutput({ code })
    expect(result).toMatchSnapshot()
  })
})
