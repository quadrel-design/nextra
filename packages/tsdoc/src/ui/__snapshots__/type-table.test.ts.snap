// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypeTable > <Banner /> 1`] = `
[
  {
    "description": "",
    "entries": [
      {
        "description": "Content of the banner.",
        "name": "children",
        "required": true,
        "tags": {},
        "type": "ReactNode",
      },
      {
        "description": "Closable banner or not.",
        "name": "dismissible",
        "required": false,
        "tags": {
          "default": "true",
        },
        "type": "boolean",
      },
      {
        "description": "Storage key to keep the banner state.",
        "name": "storageKey",
        "required": false,
        "tags": {
          "default": "'nextra-banner'",
        },
        "type": "string",
      },
    ],
    "name": "default",
  },
]
`;

exports[`TypeTable > <Callout /> 1`] = `
[
  {
    "description": "",
    "entries": [
      {
        "description": "Specifies the type of callout.
Determines the default icon if none is provided.",
        "name": "type",
        "required": false,
        "tags": {
          "default": "'default'",
        },
        "type": ""default" | "error" | "info" | "warning"",
      },
      {
        "description": "Icon displayed in the callout. Can be a string emoji or a custom React element.

Default values based on \`type\`:
- \`'💡'\` for \`type: 'default'\`
- \`'🚫'\` for \`type: 'error'\`
- \`<InformationCircleIcon />\` for \`type: 'info'\`
- \`'⚠️'\` for \`type: 'warning'\`",
        "name": "emoji",
        "required": false,
        "tags": {
          "default": "Determined by \`type\`",
        },
        "type": "ReactNode",
      },
      {
        "description": "Content to be displayed inside the callout.",
        "name": "children",
        "required": true,
        "tags": {},
        "type": "ReactNode",
      },
    ],
    "name": "default",
  },
]
`;

exports[`TypeTable > <NotFoundPage /> 1`] = `
[
  {
    "description": "",
    "entries": [
      {
        "description": "Content of the link.",
        "name": "content",
        "required": false,
        "tags": {
          "default": "'Submit an issue about broken link'",
        },
        "type": "ReactNode",
      },
      {
        "description": "Labels that can be added to the new created issue.",
        "name": "labels",
        "required": false,
        "tags": {
          "default": "'bug'",
        },
        "type": "string",
      },
      {
        "description": "Top content of page.",
        "name": "children",
        "required": false,
        "tags": {
          "default": "<H1>404: Page Not Found</H1>",
        },
        "type": "ReactNode",
      },
      {
        "description": "CSS class name.",
        "name": "className",
        "required": false,
        "tags": {},
        "type": "string",
      },
    ],
    "name": "default",
  },
]
`;

exports[`TypeTable > <Search /> 1`] = `
[
  {
    "description": "",
    "entries": [
      {
        "description": "Not found text.",
        "name": "emptyResult",
        "required": false,
        "tags": {
          "default": "'No results found.'",
        },
        "type": "ReactNode",
      },
      {
        "description": "Error text.",
        "name": "errorText",
        "required": false,
        "tags": {
          "default": "'Failed to load search index.'",
        },
        "type": "ReactNode",
      },
      {
        "description": "Loading text.",
        "name": "loading",
        "required": false,
        "tags": {
          "default": "'Loading…'",
        },
        "type": "ReactNode",
      },
      {
        "description": "Placeholder text.",
        "name": "placeholder",
        "required": false,
        "tags": {
          "default": "'Search documentation…'",
        },
        "type": "string",
      },
      {
        "description": "CSS class name.",
        "name": "className",
        "required": false,
        "tags": {},
        "type": "string",
      },
      {
        "description": "",
        "name": "searchOptions",
        "required": false,
        "tags": {},
        "type": "PagefindSearchOptions",
      },
    ],
    "name": "default",
  },
]
`;
